roofit/hs3/src/JSONFactories_HistFactory.cxx

```diff
@@ -456,73 +472,40 @@ class RooRealSumPdfFactory : public RooFit::JSONIO::Importer {
+      if (p.has_child("variables")) {
          RooJSONFactoryWSTool::getObservables(ws, p, name, observables);
          scope.setObservables(observables);
       }
       for (const auto &comp : p["samples"].children()) {
-         std::string fname(RooJSONFactoryWSTool::name(comp));
-         auto &def = comp.is_container() ? comp : p["functions"][fname.c_str()];
-         std::string fprefix = RooJSONFactoryWSTool::genPrefix(def, true);
+         std::string fname(comp["name"].val() + "_shapes");
+         std::string coefname(comp["name"].val() + "_scaleFactors");
+         std::string fprefix = RooJSONFactoryWSTool::genPrefix(comp, true);
 
          if (observables.empty()) {
             RooJSONFactoryWSTool::getObservables(ws, comp["data"], fprefix, observables);
             scope.setObservables(observables);
          }
-         if (def.has_child("overallSystematics"))
-            ::collectNames(def["overallSystematics"], sysnames);
-         if (def.has_child("histogramSystematics"))
-            ::collectNames(def["histogramSystematics"], sysnames);
-         if (def.has_child("shapeSystematics")) { // ShapeSys are special case. Create PHFs here if needed
-            std::vector<std::string> shapeSysNames;
-            ::collectNames(def["shapeSystematics"], shapeSysNames);
-            for (auto &sysname : shapeSysNames) {
-               std::string phfname = name + "_" + sysname + "_ShapeSys";
-               auto phf = scope.getObject(phfname);
-               if (!phf) {
-                  auto newphf =
-                     createPHFShapeSys(def["shapeSystematics"][sysname], phfname, ws, constraints, observables);
-                  ws.import(*newphf, RooFit::RecycleConflictNodes(), RooFit::Silence(true));
-                  scope.setObject(phfname, ws.function(phfname.c_str()));
-               }
-            }
-         }
 
-         if (comp["statError"].val_bool()) {
-            ::stackError(def["data"], sumW, sumW2);
+         for (const auto &sampleNode : p["samples"].children()) {
+            importHistSample(ws, scope, sampleNode, constraints);
          }
 
-         funcnames.push_back(fprefix + fname);
-         coefnames.push_back(fprefix + fname + "_norm");
-      }
-
-      auto phf = createPHFMCStat(name, sumW, sumW2, ws, constraints, observables, statErrorThreshold, statErrorType);
-      if (phf) {
-         ws.import(*phf, RooFit::RecycleConflictNodes(), RooFit::Silence(true));
-         scope.setObject("mcstat", ws.function(phf->GetName()));
-      }
-
-      for (const auto &sampleNode : p["samples"].children()) {
-         importHistSample(ws, scope, sampleNode);
-      }
+         auto phf = createPHFMCStat(name, sumW, sumW2, ws, constraints, observables, statErrorThreshold, statErrorType);
+         if (phf) {
+            ws.import(*phf, RooFit::RecycleConflictNodes(), RooFit::Silence(true));
+            scope.setObject("mcstat", ws.function(phf->GetName()));
```

Comment:
@cburgard, I think something might be out of order here now. In this line, you call `scope.setObject("mcstat")`, but the only time you are calling `scope.getObject("mcstat")` is in `importHistSample` that is called a few lines before! So at that time, the `"mcstat"` is not set yet, which results in crashes in the `testHS3HistFactory`.

It would be good if you could investigate this, because I don't know what the logic is really intended to be.

Reply:
Hey!
Thanks for catching this. To be honest, the whole MC stat handling probably needs to be revised anyway, seeing how we probably want this to work via 2 different methods (once with the "modifier" syntax like in pyhf, and once from the errors set on the histograms like one would do when setting up from scratch).
Hence, I didn't consciously touch the MCstat handling, also because the example I was working on doesn't actually use mcstat uncertainties.
I propose we merge this MR first and touch up the MCstat handling in a separate MR so to not overburden this one too much (it's already quite a lot of changes).

Reply:
Yes that makes sense what you say. Just wanted to make sure you're aware :+1: